gammapy/stats/li_ma.py

```diff
@@ -0,0 +1,347 @@
+"""Li and Ma algorithm to compute parameter confidence limits and significance."""
+import logging
+import numpy as np
+
+__all__ = [
+    "lm_loglikelihood",
+    "lm_dexcess_down",
+    "lm_dexcess_up",
+    "lm_dexcess",
+    "lm_significance_on_off",
+    "lm_significance"
+]
+
+log = logging.getLogger(__name__)
+
+
+def lm_loglikelihood(n_on, n_off, alpha, signal):
+    r"""Compute Li&Ma Log-likelihood.
+
+    Note: This is a little trick to avoid problems for negative signal
+    (note that this method is often called with signal=excess).
+    In that case the Poisson signal + b0 can become negative
+    and log(signal + b0) isn't defined, i.e. the likelihood cannot be computed.
+    To avoid this problem, the on and off regions are switched
+
+    For more information see :ref:`documentation <li_ma>`.
+
+    Parameters
+    ----------
+    n_on : float
+        Observed number of counts in the on region
+    n_off : float
+        Observed number of counts in the off region
+    alpha : float
+        On / off region exposure ratio for background events
+    signal : float
+        Number of excess in the on region
+
+    Returns
+    -------
+    log_likelihood : `~numpy.ndarray`
+
+    """
+    if signal > 0:
+        n1 = n_on
+        n2 = n_off
+        aa = alpha
+    else:
+        n2 = n_on
+        n1 = n_off
+        aa = 1.0 / alpha
+        signal = -aa * signal
+
+    # Maximum likelihood background estimate b0
+    tt = (1+aa) * signal - aa * (n1+n2)
+    delta = tt*tt + 4 * aa * (1+aa) * signal * n2
+    b0 = (np.sqrt(delta) - tt) / (2 + 2*aa)
+    # b0 should always be >= 0.
+    # Just to avoid possible problems with rounding errors:
+    if b0 < 0:
+        b0 = 0.
+
+    # Compute Poisson log likelihood r for two measurements: on / off
+    ll = n1 * np.log(signal + b0) + n2 * np.log(b0 / aa) - (signal + b0) - b0 / aa
+    return ll
+
+
+def lm_dexcess_down(n_on, n_off, alpha, n_sig=1.):
+    r"""Compute downward excess uncertainties of an on-off observation based on Li&Ma [1]. [LiMa1983]_
+
+    Searches the average value of the signal for which the log-likelihood ratio is n_sig away
+
+    Note: The required step in log likelihood for "nsig sigma errors" is offset = 0.5*n_sig^2
+            http://wwwasdoc.web.cern.ch/wwwasdoc/minuit/node30.html
+
+    For more information see :ref:`documentation <li_ma>`.
+
+    Parameters
+    ----------
+    n_on : array_like
+        Observed number of counts in the on region
+    n_off : array_like
+        Observed number of counts in the off region
+    alpha : array_like
+        On / off region exposure ratio for background events
+    n_sig : float
+        Number of sigma
+
+    Returns
+    -------
+    significance : `~numpy.ndarray`
+        Error estimate
+
+     References
+    ----------
+    * Li and Ma, "Analysis methods for results in gamma-ray astronomy"
+      <https://ui.adsabs.harvard.edu/abs/1983ApJ...272..317L>`_
+    """
+    scalar = False
```

Comment:
In general there is not need for this type of check. Once the expression are compatible with `~numpy.ndarray` the scalar handling come for free. See e.g.
```python3
>>> import numpy as np
>>> a = np.log(5)
>>> print(a)
1.6094379124341003
```

vs.
```python3
>>> import numpy as np
>>> a = np.log([5])
>>> print(a)
[1.60943791]
```
So just write numpy expressions...

Reply:
Agree. For the functions in poisson.py, this code supports this 2 ways of doing, using float and numpy expressions.
Should we change also poisson.py (and all the documentation) such that the functions take only numpy expression?